packages/grafana-data/src/types/events.ts

```diff
@@ -0,0 +1,181 @@
+import { IHttpResponse } from 'angular';
+import { TimeRange } from './time';
+import { DataFrame } from './dataFrame';
+
+export type AlertPayload = [string, string?];
+
+export interface ShowDashSearchPayload {
+  query?: string;
+}
+
+export interface LocationChangePayload {
+  href: string;
+}
+
+export interface ShowModalPayload {
+  model?: any;
+  modalClass?: string;
+  src?: string;
+  templateHtml?: string;
+  backdrop?: any;
+  scope?: any;
+}
+
+export interface ShowConfirmModalPayload {
+  title?: string;
+  text?: string;
+  text2?: string;
+  text2htmlBind?: boolean;
+  confirmText?: string;
+  altActionText?: string;
+  yesText?: string;
+  noText?: string;
+  icon?: string;
+
+  onConfirm?: () => void;
+  onAltAction?: () => void;
+}
+
+type DataSourceResponsePayload = IHttpResponse<any>;
+
+export interface SaveDashboardPayload {
+  overwrite?: boolean;
+  folderId?: number;
+  makeEditable?: boolean;
+}
+
+export interface PanelChangeViewPayload {
+  fullscreen?: boolean;
+  edit?: boolean;
+  panelId?: number;
+  toggle?: boolean;
+}
+
+export type RemovePanelPayload = number;
+
+export interface GraphHoverPayload {
+  pos: any;
+  panel: {
+    id: number;
+  };
+}
+
+export interface ToggleKioskModePayload {
+  exit?: boolean;
+}
+
+export interface GraphClickedPayload {
+  pos: any;
+  panel: any;
+  item: any;
+}
+
+export interface ThresholdChangedPayload {
+  threshold: any;
+  handleIndex: any;
+}
+
+export interface DashScrollPayload {
+  restore?: boolean;
+  animate?: boolean;
+  pos?: number;
+}
+
+export interface MenuElement {
+  text: string;
+  click: string;
+  role?: string;
+  shortcut?: string;
+}
+
+export interface AppEvent<T> {
+  readonly name: string;
+  payload?: T;
+}
+
+const typeList: Set<string> = new Set();
+export function eventFactory<T = undefined>(name: string): AppEvent<T> {
+  if (typeList.has(name)) {
+    throw new Error(`There is already an event defined with type '${name}'`);
+  }
+
+  typeList.add(name);
+  return { name };
+}
+
+export const alertWarning = eventFactory<AlertPayload>('alert-warning');
```

Comment:
I am not a big fan of executing functions in global scope (when bundles load) but sometimes it ok. But I wonder if we can solve the namespacing problem and that in one go. 

Probably a bad idea (not sure): 

```
export class AppEvents {
   
   alertWarning() { return eventFactory<AlertPayload>('alert-warning') }
}

Reply:
Not sure how this would work with how duplicate event names are disallowed, since invoking, say, `alertWarning` more than once would throw an error since you're basically registering the event each time you call `alertWarning`. I thought maybe memoizing calls to eventFactory would fix this but then that would stop eventFactory from throwing if someone really did accidentally try to register two events with the same name.
One (probably very bad) idea would be to do something like this:
```ts
export class AppEvents {
   
   alertWarning() {
      const alertWarningEvent = eventFactory<AlertPayload>('alert-warning');
      AppEvents.prototype.alertWarning = () => alertWarningEvent;
      return alertWarningEvent;
   }
} 
```
which would return the same value on subsequent calls, but also throw if 'alert-warning' were to be re-registered elsewhere. Probably not very good practice though

Reply:
```ts
export class AppEvents {
   
   get alertWarning() {
      const alertWarningEvent = eventFactory<AlertPayload>('alert-warning');
      Object.defineProperty(this, "alertWarning", {
        get: () => alertWarning
      };
      return alertWarningEvent;
   }
} 
```

Reply:
Maybe it's going to be easier / cleaner to declare these in global scope. If we limit the number defined in @grafana/data to only the essential ones used by plugins (alerts?), the reset in app/core. Think these will be fewer over time as we migrate more code to react

Reply:
Sounds good